LibraryPanels: Add RBAC support

kaydelaney opened the pull request:
Adds RBAC support to library panels / elements.

Fixes https://github.com/grafana/grafana/issues/50665

## Things left to do: ##

- [x] Refactor permission evaluation code

### Testing scenarios: ###

- [ ] Creating a library panel with insufficient permissions
- [ ] Reading a library panel with insufficient permissions
- [ ] Updating a library panel with insufficient permissions
- [ ] Deleting a library panel with insufficient permissions
- [ ] Ensure folder permissions are inherited

kaydelaney committed:
LibraryPanels: Add RBAC support

IevaVasiljeva commented on pkg/services/libraryelements/accesscontrol.go:
```diff
@@ -0,0 +1,21 @@
+)
+
+const (
+	ScopeLibraryPanelsRoot   = "library-panels"
```
Note: We've opted for separating other resource names that contain two words with a dot. Eg, "alert.notifications:..." instead of "alert-notifications:...". So it'd be good to follow that pattern for consistency.

kaydelaney replied:
Nice, thanks for pointing this  out

IevaVasiljeva commented on pkg/services/libraryelements/api.go:
```diff
@@ -56,7 +62,28 @@ func (l *LibraryElementService) createHandler(c *contextmodel.ReqContext) respon
+	allowed, err := l.AccessControl.Evaluate(
+		c.Req.Context(),
+		c.SignedInUser,
+		ac.EvalAny(
+			ac.EvalPermission(
+				ActionLibraryPanelsCreate,
+			),
+			ac.EvalPermission(
+				dashboards.ActionFoldersWrite,
+				dashboards.ScopeFoldersRoot+":id:"+fmt.Sprint(cmd.FolderID),
+			),
+		),
```
The access control middleware for this endpoint already checks `ActionLibraryPanelsCreate`, so this check will always pass, as user needs `ActionLibraryPanelsCreate` to get to this point. What is the intention? Do we want to make sure that they also have access to the folder? Or would it maybe make sense to scope the `ActionLibraryPanelsCreate` action to folders?

gamab commented on pkg/services/libraryelements/api.go:
```diff
@@ -15,14 +16,19 @@ import (
+		entities.Get("/", authorize(ac.EvalPermission(ActionLibraryPanelsRead)), routing.Wrap(l.getAllHandler))
+		entities.Get("/:uid", authorize(ac.EvalPermission(ActionLibraryPanelsRead, uidScope)), routing.Wrap(l.getHandler))
+		entities.Get("/:uid/connections/", authorize(ac.EvalPermission(ActionLibraryPanelsRead, uidScope)), routing.Wrap(l.getConnectionsHandler))
+		entities.Get("/name/:name", authorize(ac.EvalPermission(ActionLibraryPanelsRead, nameScope)), routing.Wrap(l.getByNameHandler))
```
I think we are missing a scope resolver registration (to translate scopes like `library-panels:name:XXX` -> `library-panels:uid:YYY` ) in order for this check to work.
I'll let @alexanderzobnin provide an example :slightly_smiling_face:

IevaVasiljeva commented on pkg/services/libraryelements/api.go:
```diff
@@ -244,7 +330,50 @@ func (l *LibraryElementService) getByNameHandler(c *contextmodel.ReqContext) res
+		c.SignedInUser,
+		ac.EvalAny(
+			ac.EvalPermission(
+				ScopeLibraryPanelsRoot+":"+action,
```
Nit: you could also use a scope provider pattern here: https://github.com/grafana/grafana/blob/main/pkg/services/accesscontrol/scope.go#L95

kaydelaney replied:
Hmm, I wanted to do this, but the library element service's `getLibraryElementByUid` method expects a pointer to a `user.SignedInUser`. Do you know if there's a way to access the currently signed-in user struct from a `ScopeAttributeResolverFunc`?

gamab commented on pkg/services/libraryelements/api.go:
```diff
@@ -122,6 +168,14 @@ func (l *LibraryElementService) getHandler(c *contextmodel.ReqContext) response.
 		return toLibraryElementError(err, "Failed to get library element")
 	}
 
+	allowed, err := l.evalPermission(c, "read", element.UID, element.FolderID)
+	if err != nil {
+		return toLibraryElementError(err, "Failed to evaluate permissions")
+	}
+	if !allowed {
+		return response.Error(http.StatusForbidden, "You do not have the permissions needed to view this library element", nil)
+	}
+
```
Same this is already handled at middleware layer by `authorize(ac.EvalPermission(ActionLibraryPanelsRead, uidScope)`
```suggestion
```

IevaVasiljeva reviewed (commented):
Great to see RBAC being extended to library panels! 🎉 

I've added some small suggestions. My main question is about how access to folders impacts access to library elements? We have several different approaches to it out there (for instance, alert actions use folder scopes while accessing annotations requires being able to access the dashboard that they belong to), all with their pros and cons.

gamab reviewed (commented):
Great to see people adding RBAC to new features :)

I only left a few comments on my first findings to help a little bit :)
I agree with @IevaVasiljeva I just wonder how libraryelements permissions should be tied to folders :+1:

mildwonkey reviewed (commented):
I can't helpfully review RBAC changes, but from a general backend platform perspective, this looks reasonable. You have more qualified folks looking at the important content of the PR, so no approvals from me, but no blockers either 😁

alexanderzobnin reviewed (dismissed):
Left some comments

alexanderzobnin reviewed (commented):
Sorry, confused buttons

IevaVasiljeva commented on pkg/services/libraryelements/api.go:
```diff
@@ -244,7 +330,50 @@ func (l *LibraryElementService) getByNameHandler(c *contextmodel.ReqContext) res
+			),
+			ac.EvalPermission(
+				dashboards.ScopeFoldersRoot+":"+action,
+				dashboards.ScopeFoldersRoot+":id:"+fmt.Sprint(folderID),
```
Alex is right, we use folder UIDs to store folder related permissions. Another option would be to use the folder guardian to check access to the folder.

kaydelaney replied:
Isn't there an ID scope resolver? https://github.com/grafana/grafana/blob/lib-rbac/pkg/services/dashboards/accesscontrol.go#L68

kaydelaney committed:
Change resource name from "library-panels" to "library.panels"

kaydelaney commented:
I've been quietly working away addressing the feedback folks have been kind enough to give, and I think I've figured out most of the things I've been stuck on, but I do still have one question:
So, it seems like when a new dashboard is created, some code is executed (`setDefaultPermissions`) that updates the permissions table for the user creating the dashboard as well as the Viewer and Editor roles if the dashboard is in the General folder. Would I be correct in assuming similar code will need to be written for the library panels use case?

Oh and since I think some people asked about it, I think library panels should more or less share the permission inheritance rules that dashboards currently do. So a library panel will inherit the permissions of the folder that contains it.

kaydelaney committed:
Clean up and finish implementation

kaydelaney commented:
Lots of changes with that last commit. I more or less just implemented everything for library elements how I saw dashboards was handled. One question I had was about how the permissions for existing library elements should be handled. Will a migration have to be written to add appropriate entries to the permissions table? @IevaVasiljeva

kaydelaney committed:
Merge remote-tracking branch 'origin/main' into lib-rbac

kaydelaney committed:
Merge remote-tracking branch 'origin/main' into lib-rbac

kaydelaney committed:
Place code behind feature flag

IevaVasiljeva commented on pkg/services/libraryelements/api.go:
```diff
@@ -2,27 +2,44 @@ package libraryelements
+			entities.Get("/name/:name", authorize(ac.EvalPermission(ActionLibraryPanelsRead, nameScope)), routing.Wrap(l.getByNameHandler))
+			entities.Patch("/:uid", authorize(ac.EvalPermission(ActionLibraryPanelsWrite, uidScope)), routing.Wrap(l.patchHandler))
+		} else {
+			entities.Post("/", routing.Wrap(l.createHandler))
```
Question: what are the current (without RBAC) safeguards for library panels? Do we require access to a folder/dashboard to CRUD a library panel?

kaydelaney replied:
Existing requirements without RBAC:
To create a library panel, a user must have edit permissions on the folder that will contain the library panel
https://github.com/grafana/grafana/blob/38c3483594080b6d7e12d94f95bcacf67c3ac368/pkg/services/libraryelements/database.go#L165

To read a library panel, if I understand the SQL correctly, a user needs view permissions on the containing folder
https://github.com/grafana/grafana/blob/38c3483594080b6d7e12d94f95bcacf67c3ac368/pkg/services/libraryelements/database.go#L271-L275

Updating a library panel also requires edit permissions on the containing folder:
https://github.com/grafana/grafana/blob/38c3483594080b6d7e12d94f95bcacf67c3ac368/pkg/services/libraryelements/database.go#L499-L509

Finally, deleting a library panel also requires edit permissions on the containing folder:
https://github.com/grafana/grafana/blob/38c3483594080b6d7e12d94f95bcacf67c3ac368/pkg/services/libraryelements/database.go#L216-L218

Let me know if you have any other questions :)

IevaVasiljeva replied:
Nice, thanks for this, this is super helpful!

So I think the options are:
1) have generic permissions to CRUD library panels + require access to the folder where the library panel resides in order to edit the library panel (this is similar to what we've done with annotations);
2) scope library panel permissions to a folder, and automatically grant library panel permissions to a user when they're set to folder viewer/editor (similarly as we've done with alerting);
3) scope library panel permissions to individual library panels + require access to the folder where the library panel resides in order to edit the library panel (this is probably unnecessarily fine-grained)

All of these approaches have some benefits and drawbacks. The current implementation seems to be mostly in line with approach 2), is that correct?

kaydelaney replied:
I think that's right. I more or less tried to copy how dashboard permissions seem to be handled, where they inherit permissions from the folder(s) they are contained in, but they can't have a greater level of access than the folder they're in. (e.g. can't have edit access to a library panel contained in a folder they only have view permissions for)

IevaVasiljeva replied:
Nice nice! I also think that approach makes the most sense.

> but they can't have a greater level of access than the folder they're in. (e.g. can't have edit access to a library panel contained in a folder they only have view permissions for)

This is an interesting one, I could see arguments in both directions. But I guess we can start with that, and see what feedback we get from the users.

IevaVasiljeva reviewed (commented):
I've added some comments, but ultimately realised that I know too little about library panels to properly review this PR. For instance, I don't know how access to library panels is protected currently. Could we maybe go through this PR together in a call? Alternatively, if there are any design docs for access to library panels, that might also help me.

IevaVasiljeva commented:
It might also be helpful to split this PR into smaller parts. For instance, to tackle read access to library panels first, and then do creating/editing/deleting.

kaydelaney commented on pkg/api/accesscontrol.go:
```diff
@@ -408,6 +410,48 @@ func (hs *HTTPServer) declareFixedRoles() error {
+			Name:        "fixed:library.panels:writer",
+			DisplayName: "Library panel writer",
+			Group:       "Library panels",
+			Description: "Create, read, write or delete all library panels and their permissions.",
```
the permissions are being concat-ed to the reader role permissions
```go
			Permissions: ac.ConcatPermissions(libraryPanelsReaderRole.Role.Permissions, []ac.Permission{
				{Action: libraryelements.ActionLibraryPanelsWrite, Scope: libraryelements.ScopeLibraryPanelsAll},
				{Action: libraryelements.ActionLibraryPanelsDelete, Scope: libraryelements.ScopeLibraryPanelsAll},
				{Action: libraryelements.ActionLibraryPanelsCreate, Scope: libraryelements.ScopeLibraryPanelsAll},
			}),
```

IevaVasiljeva replied:
Ah yeah, I misread this 🙈 Thanks for clarifying!

kaydelaney commented on pkg/services/libraryelements/api.go:
```diff
@@ -2,27 +2,44 @@ package libraryelements
+			entities.Get("/", authorize(ac.EvalPermission(ActionLibraryPanelsRead)), routing.Wrap(l.getAllHandler))
+			entities.Get("/:uid", authorize(ac.EvalPermission(ActionLibraryPanelsRead, uidScope)), routing.Wrap(l.getHandler))
+			entities.Get("/:uid/connections/", authorize(ac.EvalPermission(ActionLibraryPanelsRead, uidScope)), routing.Wrap(l.getConnectionsHandler))
+			entities.Get("/name/:name", authorize(ac.EvalPermission(ActionLibraryPanelsRead, nameScope)), routing.Wrap(l.getByNameHandler))
```
That makes sense :+1:

kaydelaney commented on pkg/services/accesscontrol/ossaccesscontrol/permissions_services.go:
```diff
@@ -186,6 +190,83 @@ func ProvideDashboardPermissions(
+var LibraryElementEditActions = append(LibraryElementViewActions, []string{libraryelements.ActionLibraryPanelsWrite, libraryelements.ActionLibraryPanelsDelete}...)
+var LibraryElementAdminActions = LibraryElementEditActions
+
+func ProvideLibraryElementPermissions(
```
added to LibraryElementService in libraryelements.go:
```go
func ProvideService(cfg *setting.Cfg, sqlStore db.DB, routeRegister routing.RouteRegister, folderService folder.Service, features featuremgmt.FeatureToggles, ac accesscontrol.AccessControl, libraryElementPermissionsService accesscontrol.LibraryElementPermissionsService) *LibraryElementService {
	l := &LibraryElementService{
		Cfg:                       cfg,
		SQLStore:                  sqlStore,
		RouteRegister:             routeRegister,
		folderService:             folderService,
		log:                       log.New("library-elements"),
		features:                  features,
		AccessControl:             ac,
		libraryElementPermissions: libraryElementPermissionsService,
	}

	l.registerAPIEndpoints()
	ac.RegisterScopeAttributeResolver(LibraryPanelUIDScopeResolver(l, l.folderService))

	return l
}
```

and then used as part of `updatePermissions` in database.go (second last line)

```go
func (l *LibraryElementService) updatePermissions(c context.Context, model *model.LibraryElementDTO, signedInUser identity.Requester) error {
	inFolder := model.FolderID > 0
	var permissions []accesscontrol.SetResourcePermissionCommand
	namespaceID, userIDstr := signedInUser.GetNamespacedID()
	userID, err := identity.IntIdentifier(namespaceID, userIDstr)
	if err != nil {
		return err
	}

	if namespaceID == identity.NamespaceUser {
		permissions = append(permissions, accesscontrol.SetResourcePermissionCommand{
			UserID: userID, Permission: dashboards.PERMISSION_ADMIN.String(),
		})
	}

	if !inFolder {
		permissions = append(permissions, []accesscontrol.SetResourcePermissionCommand{
			{BuiltinRole: string(org.RoleEditor), Permission: dashboards.PERMISSION_EDIT.String()},
			{BuiltinRole: string(org.RoleViewer), Permission: dashboards.PERMISSION_VIEW.String()},
		}...)
	}

	_, err = l.libraryElementPermissions.SetPermissions(c, model.OrgID, model.UID, permissions...)
	return err
}
```

IevaVasiljeva replied:
Thanks for linking the usage. The way we've used the permission service framework so far is for resources that have a frontend component that allows setting permissions on them through the UI (like [this tab](https://ops.grafana-ops.net/d/d8e4202e-6737-4ec2-b62d-c15de841663a/api-key-deprecation?orgId=1&editview=permissions) for dashboards). So I feel like it might be an overkill for library elements.

kaydelaney replied:
That seems reasonable

kaydelaney commented on pkg/services/accesscontrol/ossaccesscontrol/permissions_services.go:
```diff
@@ -231,9 +312,9 @@ func ProvideFolderPermissions(
-			"View":  append(DashboardViewActions, FolderViewActions...),
-			"Edit":  append(DashboardEditActions, FolderEditActions...),
-			"Admin": append(DashboardAdminActions, FolderAdminActions...),
+			"View":  append(DashboardViewActions, append(LibraryElementViewActions, FolderViewActions...)...),
+			"Edit":  append(DashboardEditActions, append(LibraryElementEditActions, FolderEditActions...)...),
+			"Admin": append(DashboardAdminActions, append(LibraryElementAdminActions, FolderAdminActions...)...),
```
Yes, I think I left a comment asking about that a while ago. Do you know of any existing migration I could base this off?

IevaVasiljeva replied:
We did a similar migration for alert actions: https://github.com/grafana/grafana/blob/main/pkg/services/sqlstore/migrations/accesscontrol/dashboard_permissions.go#L333C34-L333C34

kaydelaney commented on pkg/api/accesscontrol.go:
```diff
@@ -408,6 +410,48 @@ func (hs *HTTPServer) declareFixedRoles() error {
+			Description: "Create library panel in general folder.",
+			Group:       "Library panels",
+			Permissions: []ac.Permission{
+				{Action: dashboards.ActionFoldersRead, Scope: dashboards.ScopeFoldersProvider.GetResourceScopeUID(ac.GeneralFolderUID)},
```
Was more or less just copying what's done for the dashboardsCreatorRole:
```go
	dashboardsCreatorRole := ac.RoleRegistration{
		Role: ac.RoleDTO{
			Name:        "fixed:dashboards:creator",
			DisplayName: "Dashboard creator",
			Description: "Create dashboard in general folder.",
			Group:       "Dashboards",
			Permissions: []ac.Permission{
				{Action: dashboards.ActionFoldersRead, Scope: dashboards.ScopeFoldersProvider.GetResourceScopeUID(ac.GeneralFolderUID)},
				{Action: dashboards.ActionDashboardsCreate, Scope: dashboards.ScopeFoldersProvider.GetResourceScopeUID(ac.GeneralFolderUID)},
			},
		},
		Grants: []string{"Editor"},
	}
```

kaydelaney committed:
Remove unneeded name scope resolver

IevaVasiljeva commented on pkg/services/libraryelements/accesscontrol.go:
```diff
@@ -0,0 +1,69 @@
+		if err != nil {
+			return nil, err
+		}
+		return append(inheritedScopes, dashboards.ScopeFoldersProvider.GetResourceScopeUID(libElDTO.FolderUID), ScopeLibraryPanelsProvider.GetResourceScopeUID(uid)), nil
```
I don't think we can mix UID scopes for folder UIDs and library element UID itself, as a library element can probably have the same UID as a folder? Maybe it's enough to only have folder UID based scopes? It would mean that a user can either CRUD all or no library panels in a given folder, do you think that would be fine-grained enough?

kaydelaney replied:
Hmm, it seems like the same thing is being done for dashboards though, unless I'm missing something?

https://github.com/grafana/grafana/blob/03585279e956ac5ae2f1afe7ec5c69952598a0ee/pkg/services/dashboards/accesscontrol.go#L188-L193

IevaVasiljeva commented on pkg/services/libraryelements/api.go:
```diff
@@ -56,7 +72,25 @@ func (l *LibraryElementService) createHandler(c *contextmodel.ReqContext) respon
 	}
-	element, err := l.CreateElement(c.Req.Context(), c.SignedInUser, cmd)
+
+	if l.features.IsEnabled(featuremgmt.FlagLibraryPanelRBAC) {
+		allowed, err := l.AccessControl.Evaluate(
+			c.Req.Context(),
+			c.SignedInUser,
+			ac.EvalPermission(
+				dashboards.ActionFoldersWrite,
+				dashboards.ScopeFoldersRoot+":id:"+fmt.Sprint(cmd.FolderID),
+			),
+		)
+		if err != nil {
+			return toLibraryElementError(err, "Failed to evaluate permissions")
+		}
+		if !allowed {
+			return response.Error(http.StatusForbidden, "You do not have the permissions needed to create a library element in this folder", nil)
+		}
+	}
```
We still evaluate access to the folder from [here](https://github.com/grafana/grafana/blob/38c3483594080b6d7e12d94f95bcacf67c3ac368/pkg/services/libraryelements/database.go#L165), right? I think that is probably enough, and this additional check is not needed.

IevaVasiljeva commented on pkg/services/libraryelements/database.go:
```diff
@@ -202,14 +203,46 @@ func (l *LibraryElementService) createLibraryElement(c context.Context, signedIn
 	return dto, err
 }
 
+func (l *LibraryElementService) updatePermissions(c context.Context, model *model.LibraryElementDTO, signedInUser identity.Requester) error {
```
This would be needed if we want to allow setting permissions on an individual library panel level. If we're fine with only granting permissions on a folder level (either you can or can't CRUD library elements in a given folder), then this would not be necessary. What do you think?

kaydelaney replied:
Not actually sure quite how fine-grained we need things to be. I'll try and get an answer for you.

kaydelaney replied:
Got some clarification and we do actually want permissions to be at the individual library panel level

IevaVasiljeva reviewed (commented):
Thanks for all the responses! I've added some more comments. I think the permission setup that you're going for is closer to what alerting has than what dashboards have, so it's worth taking a look at that.

I'm also happy to chat about this in a call if that would be helpful :)

kaydelaney committed:
Merge remote-tracking branch 'origin/main' into lib-rbac

kaydelaney committed:
Remove redundant code

IevaVasiljeva reviewed (commented):
Great work!

I'd be happy to approve this PR, as long as we make the following changes in this or a follow-up PR:
* add tests;
* adding name scope resolver for that one endpoint that allows getting a library panel by name (or am I overlooking it?);
* add a migration to add the new library panel permissions to folder Viewers, Editors and Admins (similar to [this](https://github.com/grafana/grafana/blob/main/pkg/services/sqlstore/migrations/accesscontrol/dashboard_permissions.go#L333C34-L333C34)) - otherwise users who can currently use library panels won't be able to use them anymore;
* frontend changes (for instance, to make sure that users who can't delete a library panel don't see the delete button etc) - not sure how many changes are necessary there.

Changes that I think we should make, but are ultimately up to your squad to make a call on:
* as library panel actions can be scoped to folders, I think we can remove folder permission checks. So, for instance, if a user has _library.panels:delete_ with scope _folders:uid:my_folder_, then they can delete library panels in _my_folder_ without being able to edit the folder itself;
* don't create a permission service and don't set permissions to library panel creator.

IevaVasiljeva commented on pkg/services/libraryelements/api.go:
```diff
@@ -2,27 +2,43 @@ package libraryelements
+			entities.Get("/", authorize(ac.EvalPermission(ActionLibraryPanelsRead)), routing.Wrap(l.getAllHandler))
+			entities.Get("/:uid", authorize(ac.EvalPermission(ActionLibraryPanelsRead, uidScope)), routing.Wrap(l.getHandler))
+			entities.Get("/:uid/connections/", authorize(ac.EvalPermission(ActionLibraryPanelsRead, uidScope)), routing.Wrap(l.getConnectionsHandler))
+			entities.Get("/name/:name", authorize(ac.EvalPermission(ActionLibraryPanelsRead)), routing.Wrap(l.getByNameHandler))
```
One more thing - it seems like we're currently checking only the general permission to list any library folder here (as we're omitting the scope). Do we need to add a name scope resolver for this? I see that you've added `ErrElementNameNotUnique`, so you probably thought of this already :)

kaydelaney commented on pkg/services/libraryelements/database.go:
```diff
@@ -202,14 +203,46 @@ func (l *LibraryElementService) createLibraryElement(c context.Context, signedIn
+		})
+	}
+
+	if !inFolder {
```
They can be in the "General" folder, or an actual folder

IevaVasiljeva replied:
Interesting, I think I missed it, as I don't see the panels created in the general folder through the UI.

Do all the viewers get view access to library panels in the general folder, and editors and admins edit access? If so, these permissions will need to be assigned to the Viewer/Editor roles (very similar to what you've done [here](https://github.com/grafana/grafana/pull/73475/files/212ba2e4bd90b809d7b13df26fa4c2b9a3e96762#diff-899e7e61e0a70e0db3ae79892147298ebf5030e05469ea529dc69641f4a7aa7cR413), but for viewing, writing and deleting).

kaydelaney replied:
I think that would  make sense, yes. I'll make the change

kaydelaney committed:
Remove permissions service

kaydelaney committed:
Assign roles for reading/writing library panels in general folder to Viewer/Editor roles

kaydelaney commented:
> Changes that I think we should make, but are ultimately up to your squad to make a call on:
> 
>     * as library panel actions can be scoped to folders, I think we can remove folder permission checks. So, for instance, if a user has _library.panels:delete_ with scope _folders:uid:my_folder_, then they can delete library panels in _my_folder_ without being able to edit the folder itself;
> 
>     * don't create a permission service and don't set permissions to library panel creator.

Hmm, so when we are creating a library panel, we check if  the user has the `ActionLibraryPanelsCreate` permission scoped to the folder they want to create the panel in? Would that not require entries in the `permission` table? How do we do that without the permission service? I feel like I'm misunderstanding something :sweat_smile:

IevaVasiljeva commented:
> @IevaVasiljeva
> 
> > Changes that I think we should make, but are ultimately up to your squad to make a call on:
> > ```
> > * as library panel actions can be scoped to folders, I think we can remove folder permission checks. So, for instance, if a user has _library.panels:delete_ with scope _folders:uid:my_folder_, then they can delete library panels in _my_folder_ without being able to edit the folder itself;
> > 
> > * don't create a permission service and don't set permissions to library panel creator.
> > ```
> 
> Hmm, so when we are creating a library panel, we check if the user has the `ActionLibraryPanelsCreate` permission scoped to the folder they want to create the panel in? Would that not require entries in the `permission` table? How do we do that without the permission service? I feel like I'm misunderstanding something 😅

We could do what we're doing with alerting, and add library panel permissions to the [folder permission service](https://github.com/grafana/grafana/blob/main/pkg/services/accesscontrol/ossaccesscontrol/permissions_services.go#L194-L203). So when someone is granted the permission to view a folder, they'd also get a permission to view the library panels in it. And when someone can edit the folder, they could also edit/create/delete the library panels. The downside is that it would be hard to give someone permissions to view the folder **without** allowing them to see the library panels, but we're actually working on a project that would mitigate this issue (I'll send you the product DNA in Slack).

But I really don't want to block this work on that. So I'll go ahead and approve this PR, and we can decide later if we want to include library panel permissions in the folder permissions service or not.

IevaVasiljeva reviewed (approved):
Great work!

Approving this PR, as I imagine it might be easier to tackle any [follow-up work](https://github.com/grafana/grafana/pull/73475#pullrequestreview-1659342991) in separate PRs, as this one has gotten pretty huge already :)

kaydelaney committed:
Remove unnecessary permission eval on /name/:name api

kaydelaney committed:
Add appropriate library panel permissions to folder permissions service

kaydelaney committed:
Merge remote-tracking branch 'origin/main' into lib-rbac

kaydelaney commented:
Thanks! I needed to merge anyway, so I added that change to the folder permissions service while I was at it.
Will follow up with some tests in another PR, though I might need some help with that. Tried writing some earlier without much luck

kaydelaney committed:
Add library panel actions folder migration

kaydelaney committed:
Fix failing tests

kaydelaney committed:
Fix linting issue
